pkg/envoy/xds/ack.go

```diff
@@ -112,19 +118,50 @@ func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName strin
 
 	version, _ := m.mutator.Upsert(typeURL, resourceName, resource, true)
 
-	comp := &pendingCompletion{
-		completion:              completion,
-		version:                 version,
-		typeURL:                 typeURL,
-		remainingNodesResources: make(map[string]map[string]struct{}, len(nodeIDs)),
-	}
+	comp, found := m.pendingCompletions[completion]
+	if found {
+		if comp.typeURL != typeURL {
+			log.WithFields(logrus.Fields{
+				logfields.XDSTypeURL: typeURL,
+			}).Fatalf("Completion reused with mismatching typeURL (have %s)",
+				comp.typeURL)
+		}
+		if len(nodeIDs) != len(comp.remainingNodesResources) {
+			log.WithFields(logrus.Fields{
+				logfields.XDSTypeURL: typeURL,
+			}).Fatalf("Completion reused with mismatching nodeIDs: %v (have %v)",
+				nodeIDs, comp.remainingNodesResources)
+		}
+		for _, nodeID := range nodeIDs {
+			if resMap, ok := comp.remainingNodesResources[nodeID]; !ok {
+				log.WithFields(logrus.Fields{
+					logfields.XDSTypeURL: typeURL,
+				}).Fatalf("Completion reused with mismatching nodeID: %s",
+					nodeID)
+			} else if _, ok := resMap[resourceName]; !ok {
+				log.WithFields(logrus.Fields{
+					logfields.XDSTypeURL: typeURL,
+				}).Fatalf("Completion reused with mismatching resourceName: %s",
+					resourceName)
+			}
+		}
+		// Update the version in the existing pendingCompletion
+		comp.version = version
+	} else {
+		// Create a new pendingCompletion
+		comp = &pendingCompletion{
+			completion:              completion,
```

Comment:
If you move that into the key of your new map, remove it from the `pendingCompletion` struct altogether.

Reply:
Right, will do!